File: tools/cuda_to_uptk/generate_map.py
#!/usr/bin/env python3
"""Generate CUDA->UPTK API mappings from uptk source convert files."""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
RUNTIME_CONVERT = ROOT / "src" / "runtime" / "runtime_fun_convert.cpp"
DRIVER_CONVERT = ROOT / "src" / "driver" / "driver_fun_convert.cpp"
OUTPUT = Path(__file__).resolve().parent / "api_map.json"

RUNTIME_DEF = re.compile(
    r"^\s*(?:__host__\s+)?UPTKError(?:_t)?\s+(UPTK[A-Za-z0-9_]+)\s*\(",
    re.MULTILINE,
)
RUNTIME_CALL = re.compile(r"\bcuda([A-Za-z0-9_]+)\s*\(")

DRIVER_DEF = re.compile(
    r"^\s*UPTKError\s+(UP[A-Za-z0-9_]+)\s*\(",
    re.MULTILINE,
)
DRIVER_CALL = re.compile(r"\bcu([A-Za-z0-9_]+)\s*\(")

DRIVER_TYPE_RULES = {
    "CUresult": "UPTKError",
    "CUDA_SUCCESS": "UPTKSuccess",
    "CUdeviceptr": "UPTKdeviceptr",
    "CUdeviceptr_v2": "UPTKdeviceptr",
}


def _collect_pairs(source: Path, def_pattern: re.Pattern, call_pattern: re.Pattern, uptk_prefix: str, cuda_prefix: str) -> dict[str, str]:
    text = source.read_text(encoding="utf-8")
    mapping: dict[str, str] = {}
    fallbacks: dict[str, str] = {}

    for match in def_pattern.finditer(text):
        uptk_name = match.group(1)
        block_end = text.find("\n}", match.end())
        if block_end == -1:
            continue
        block = text[match.start() : block_end]
        calls = call_pattern.findall(block)
        if not calls:
            continue
        cuda_name = f"{cuda_prefix}{calls[0]}"
        canonical = f"{uptk_prefix}{cuda_name[len(cuda_prefix):]}"
        if uptk_name == canonical:
            mapping[cuda_name] = uptk_name
        else:
            fallbacks.setdefault(cuda_name, uptk_name)

    for cuda_name, uptk_name in fallbacks.items():
        mapping.setdefault(cuda_name, uptk_name)

    return mapping


HEADER_NAME_SUFFIXES = {
    "_runtime_api",
    "_runtime",
    "_driver_types",
    "_library_types",
}


def _is_header_name(cuda_token: str) -> bool:
    return any(cuda_token.endswith(suffix) for suffix in HEADER_NAME_SUFFIXES)


def _collect_runtime_types(runtime_functions: dict[str, str]) -> dict[str, str]:
    header = (ROOT / "include" / "UPTK_runtime_api.h").read_text(encoding="utf-8")
    mapping: dict[str, str] = {}
    for token in re.findall(r"\bUPTK[A-Za-z0-9_]+\b", header):
        if token.startswith("UPTK"):
            cuda_token = "cuda" + token[4:]
            if _is_header_name(cuda_token):
                continue
            if cuda_token in runtime_functions:
                continue
            mapping[cuda_token] = token
    return mapping


def _collect_driver_types() -> dict[str, str]:
    header = (ROOT / "include" / "UPTK.h").read_text(encoding="utf-8")
    mapping = dict(DRIVER_TYPE_RULES)
    for token in re.findall(r"\bUPTK[A-Za-z0-9_]+\b", header):
        if token.startswith("UPTK"):
            cuda_token = "CU" + token[4:]
            mapping[cuda_token] = token
    return mapping


def generate() -> dict:
    runtime_functions = _collect_pairs(RUNTIME_CONVERT, RUNTIME_DEF, RUNTIME_CALL, "UPTK", "cuda")
    driver_functions = _collect_pairs(DRIVER_CONVERT, DRIVER_DEF, DRIVER_CALL, "UP", "cu")

    # Ensure common runtime calls exist even if wrapper naming differs.
    for cuda_name in list(runtime_functions):
        simple = "UPTK" + cuda_name[4:]
        runtime_functions.setdefault(cuda_name, simple)

    for cu_name in list(driver_functions):
        simple = "UP" + cu_name[2:]
        driver_functions.setdefault(cu_name, simple)

    return {
        "runtime_functions": dict(sorted(runtime_functions.items(), key=lambda item: len(item[0]), reverse=True)),
        "driver_functions": dict(sorted(driver_functions.items(), key=lambda item: len(item[0]), reverse=True)),
        "runtime_types": dict(sorted(_collect_runtime_types(runtime_functions).items(), key=lambda item: len(item[0]), reverse=True)),
        "driver_types": dict(sorted(_collect_driver_types().items(), key=lambda item: len(item[0]), reverse=True)),
    }


def main() -> None:
    data = generate()
    OUTPUT.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(f"Wrote {OUTPUT}")
    print(
        f"runtime_functions={len(data['runtime_functions'])}, "
        f"driver_functions={len(data['driver_functions'])}, "
        f"runtime_types={len(data['runtime_types'])}, "
        f"driver_types={len(data['driver_types'])}"
    )


if __name__ == "__main__":
    main()

File: tools/cuda_to_uptk/generate_map.js
#!/usr/bin/env node
/** Generate CUDA->UPTK API mappings from uptk source convert files. */

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..", "..");
const RUNTIME_CONVERT = path.join(ROOT, "src", "runtime", "runtime_fun_convert.cpp");
const DRIVER_CONVERT = path.join(ROOT, "src", "driver", "driver_fun_convert.cpp");
const OUTPUT = path.join(__dirname, "api_map.json");

const RUNTIME_DEF = /^\s*(?:__host__\s+)?UPTKError(?:_t)?\s+(UPTK[A-Za-z0-9_]+)\s*\(/gm;
const RUNTIME_CALL = /\bcuda([A-Za-z0-9_]+)\s*\(/g;
const DRIVER_DEF = /^\s*UPTKError\s+(UP[A-Za-z0-9_]+)\s*\(/gm;
const DRIVER_CALL = /\bcu([A-Za-z0-9_]+)\s*\(/g;

const DRIVER_TYPE_RULES = {
  CUresult: "UPTKError",
  CUDA_SUCCESS: "UPTKSuccess",
  CUdeviceptr: "UPTKdeviceptr",
  CUdeviceptr_v2: "UPTKdeviceptr",
};

function collectPairs(sourcePath, defPattern, callPattern, uptkPrefix, cudaPrefix) {
  const text = fs.readFileSync(sourcePath, "utf8");
  const mapping = {};
  const fallbacks = {};

  for (const match of text.matchAll(defPattern)) {
    const uptkName = match[1];
    const blockEnd = text.indexOf("\n}", match.index);
    if (blockEnd === -1) continue;
    const block = text.slice(match.index, blockEnd);
    const calls = [...block.matchAll(callPattern)].map((m) => m[1]);
    if (!calls.length) continue;
    const cudaName = `${cudaPrefix}${calls[0]}`;
    const canonical = `${uptkPrefix}${cudaName.slice(cudaPrefix.length)}`;
    if (uptkName === canonical) mapping[cudaName] = uptkName;
    else if (!(cudaName in fallbacks)) fallbacks[cudaName] = uptkName;
  }

  for (const [cudaName, uptkName] of Object.entries(fallbacks)) {
    if (!(cudaName in mapping)) mapping[cudaName] = uptkName;
  }

  return mapping;
}

const HEADER_NAME_SUFFIXES = ["_runtime_api", "_runtime", "_driver_types", "_library_types"];

function isHeaderName(cudaToken) {
  return HEADER_NAME_SUFFIXES.some((suffix) => cudaToken.endsWith(suffix));
}

function collectRuntimeTypes(runtimeFunctions) {
  const header = fs.readFileSync(path.join(ROOT, "include", "UPTK_runtime_api.h"), "utf8");
  const mapping = {};
  for (const token of header.match(/\bUPTK[A-Za-z0-9_]+\b/g) || []) {
    const cudaToken = `cuda${token.slice(4)}`;
    if (isHeaderName(cudaToken)) continue;
    if (cudaToken in runtimeFunctions) continue;
    mapping[cudaToken] = token;
  }
  return mapping;
}

function collectDriverTypes() {
  const header = fs.readFileSync(path.join(ROOT, "include", "UPTK.h"), "utf8");
  const mapping = { ...DRIVER_TYPE_RULES };
  for (const token of header.match(/\bUPTK[A-Za-z0-9_]+\b/g) || []) {
    mapping[`CU${token.slice(4)}`] = token;
  }
  return mapping;
}

function sortByKeyLengthDesc(obj) {
  return Object.fromEntries(
    Object.entries(obj).sort((a, b) => b[0].length - a[0].length)
  );
}

function generate() {
  const runtimeFunctions = collectPairs(
    RUNTIME_CONVERT,
    RUNTIME_DEF,
    RUNTIME_CALL,
    "UPTK",
    "cuda"
  );
  const driverFunctions = collectPairs(
    DRIVER_CONVERT,
    DRIVER_DEF,
    DRIVER_CALL,
    "UP",
    "cu"
  );

  for (const cudaName of Object.keys(runtimeFunctions)) {
    runtimeFunctions[cudaName] ||= `UPTK${cudaName.slice(4)}`;
  }
  for (const cuName of Object.keys(driverFunctions)) {
    driverFunctions[cuName] ||= `UP${cuName.slice(2)}`;
  }

  return {
    runtime_functions: sortByKeyLengthDesc(runtimeFunctions),
    driver_functions: sortByKeyLengthDesc(driverFunctions),
    runtime_types: sortByKeyLengthDesc(collectRuntimeTypes(runtimeFunctions)),
    driver_types: sortByKeyLengthDesc(collectDriverTypes()),
  };
}

const data = generate();
fs.writeFileSync(OUTPUT, `${JSON.stringify(data, null, 2)}\n`, "utf8");
console.log(`Wrote ${OUTPUT}`);
console.log(
  `runtime_functions=${Object.keys(data.runtime_functions).length}, ` +
    `driver_functions=${Object.keys(data.driver_functions).length}, ` +
    `runtime_types=${Object.keys(data.runtime_types).length}, ` +
    `driver_types=${Object.keys(data.driver_types).length}`
);
